File: packages/simpod_core/lib/src/models/ax_node.dart
class AXNode {
  const AXNode({
    required this.role,
    this.subrole,
    this.label,
    this.value,
    this.identifier,
    this.title,
    this.help,
    required this.frameX,
    required this.frameY,
    required this.frameWidth,
    required this.frameHeight,
    this.enabled = true,
    this.focused = false,
    this.hidden = false,
    this.children = const [],
    this.depth = 0,
  });

  factory AXNode.fromJson(Map<String, dynamic> json, {int depth = 0}) {
    final frame = json['frame'] as Map<String, dynamic>? ?? {};
    final childrenJson = json['children'] as List<dynamic>? ?? [];
    return AXNode(
      role: json['role'] as String? ?? 'AXUnknown',
      subrole: json['subrole'] as String?,
      label: json['label'] as String?,
      value: json['value']?.toString(),
      identifier: json['identifier'] as String?,
      title: json['title'] as String?,
      help: json['help'] as String?,
      frameX: (frame['x'] as num?)?.toDouble() ?? 0,
      frameY: (frame['y'] as num?)?.toDouble() ?? 0,
      frameWidth: (frame['width'] as num?)?.toDouble() ?? 0,
      frameHeight: (frame['height'] as num?)?.toDouble() ?? 0,
      enabled: json['enabled'] as bool? ?? true,
      focused: json['focused'] as bool? ?? false,
      hidden: json['hidden'] as bool? ?? false,
      children: childrenJson
          .whereType<Map<String, dynamic>>()
          .map((c) => AXNode.fromJson(c, depth: depth + 1))
          .toList(),
      depth: depth,
    );
  }

  static AXNode? fromApiResponse(List<dynamic> json) {
    if (json.isEmpty) return null;
    return AXNode.fromJson(json[0] as Map<String, dynamic>);
  }

  final String role;
  final String? subrole;
  final String? label;
  final String? value;
  final String? identifier;
  final String? title;
  final String? help;
  final double frameX;
  final double frameY;
  final double frameWidth;
  final double frameHeight;
  final bool enabled;
  final bool focused;
  final bool hidden;
  final List<AXNode> children;

  /// Depth in the tree (0 = root). Set during parsing to simplify
  /// indent calculations in the tree view.
  final int depth;

  String get displayName {
    return title ?? label ?? value ?? identifier ?? '';
  }

  String get shortRole {
    if (role.startsWith('AX')) return role.substring(2);
    return role;
  }

  bool get hasFrame => frameWidth > 0 && frameHeight > 0;

  int get totalCount => 1 + children.fold(0, (sum, c) => sum + c.totalCount);

  List<AXNode> flatten() {
    final result = <AXNode>[this];
    for (final child in children) {
      result.addAll(child.flatten());
    }
    return result;
  }

  AXNode? hitTest(double x, double y) {
    if (!_contains(x, y)) return null;
    for (final child in children) {
      final hit = child.hitTest(x, y);
      if (hit != null) return hit;
    }
    return this;
  }

  bool _contains(double x, double y) {
    return x >= frameX &&
        x < frameX + frameWidth &&
        y >= frameY &&
        y < frameY + frameHeight;
  }

  @override
  String toString() =>
      'ViewNode($shortRole ${displayName.isNotEmpty ? '"$displayName"' : ''})';

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is AXNode &&
          other.role == role &&
          other.subrole == subrole &&
          other.label == label &&
          other.value == value &&
          other.identifier == identifier &&
          other.title == title &&
          other.help == help &&
          other.frameX == frameX &&
          other.frameY == frameY &&
          other.frameWidth == frameWidth &&
          other.frameHeight == frameHeight &&
          other.enabled == enabled &&
          other.focused == focused &&
          other.hidden == hidden &&
          other.children == children &&
          other.depth == depth;

  @override
  int get hashCode =>
      role.hashCode ^
      subrole.hashCode ^
      label.hashCode ^
      value.hashCode ^
      identifier.hashCode ^
      title.hashCode ^
      help.hashCode ^
      frameX.hashCode ^
      frameY.hashCode ^
      frameWidth.hashCode ^
      frameHeight.hashCode ^
      enabled.hashCode ^
      focused.hashCode ^
      hidden.hashCode ^
      children.hashCode ^
      depth.hashCode;
}

File: packages/native/Sources/SimpodHelper/Accessiblity/AXNode.swift
//
//  AXNode.swift
//  SimpodHelper
//
//  Created by Tomiwa Idowu on 6/9/26.
//
//  Standalone value type for the simulator UI snapshot, plus the
//  hit-test / leaf-classification / dedup-merge helpers that operate
//  on it. Pulled out of the bridge file so the recursive maths
//  isn't tangled with the AXP runtime plumbing.
//

import CoreGraphics
import Foundation

/// A snapshot of one element in the simulator's on-screen UI tree.
struct AXNode: Equatable, Sendable {
    let role: String
    let subrole: String?
    let label: String?
    let value: String?
    let identifier: String?
    let title: String?
    let help: String?
    let frameX: Double
    let frameY: Double
    let frameWidth: Double
    let frameHeight: Double
    let enabled: Bool
    let focused: Bool
    let hidden: Bool
    let children: [AXNode]
}

// MARK: - JSON projection

extension AXNode {
    /// Recursive dictionary matching the JSON schema. Sorted keys
    /// keep diffs and snapshot tests stable.
    var dictionary: [String: Any] {
        [
            "role":       role,
            "subrole":    subrole as Any? ?? NSNull(),
            "label":      label as Any? ?? NSNull(),
            "value":      value as Any? ?? NSNull(),
            "identifier": identifier as Any? ?? NSNull(),
            "title":      title as Any? ?? NSNull(),
            "help":       help as Any? ?? NSNull(),
            "frame": [
                "x":      frameX,
                "y":      frameY,
                "width":  frameWidth,
                "height": frameHeight,
            ],
            "enabled":  enabled,
            "focused":  focused,
            "hidden":   hidden,
            "children": children.map(\.dictionary),
        ]
    }

    /// Serialized JSON bytes, ready to return from an HTTP handler.
    var jsonData: Data {
        // Top-level array wrapping, matching `axe describe-ui` shape.
        let payload: [Any] = [dictionary]
        return (try? JSONSerialization.data(
            withJSONObject: payload, options: [.sortedKeys]
        )) ?? Data()
    }
}

// MARK: - Client-side hit-test

extension AXNode {
    /// Deepest descendant whose frame contains `(px, py)`, or `self`
    /// if no child claims the point. `nil` when the point is outside
    /// `self.frame` entirely.
    func hitTest(px: Double, py: Double) -> AXNode? {
        guard containsPoint(px, py) else { return nil }
        for child in children {
            if let hit = child.hitTest(px: px, py: py) { return hit }
        }
        return self
    }

    fileprivate func containsPoint(_ px: Double, _ py: Double) -> Bool {
        px >= frameX && px < frameX + frameWidth &&
        py >= frameY && py < frameY + frameHeight
    }
}

// MARK: - Leaf / container classification

extension AXNode {
    /// Roles that fully describe themselves as a single tappable
    /// element. A childless node with one of these roles is a
    /// genuine leaf — the grid sweep can skip its interior.
    /// Everything else (groups, scroll areas, toolbars, …) is
    /// treated as a *container* even when childless, because iOS
    /// frequently hides children from the recursive walk but
    /// still exposes them to positional hit-tests.
    static let terminalRoles: Set<String> = [
        "AXStaticText", "AXButton", "AXImage",
        "AXTextField", "AXTextArea", "AXSecureTextField",
        "AXLink", "AXCheckBox", "AXRadioButton",
        "AXSlider", "AXSwitch", "AXStepper",
        "AXValueIndicator", "AXPopUpButton",
        "AXMenuItem", "AXMenuButton",
        "AXDisclosureTriangle", "AXProgressIndicator",
    ]

    /// `true` when this node is a childless terminal element whose
    /// interior the grid sweep can safely skip.
    var isTerminal: Bool {
        children.isEmpty && Self.terminalRoles.contains(role)
    }

    /// Collects frames of genuine terminal leaves. The grid sweep
    /// skips points inside these — they're already fully described.
    /// Container frames are deliberately *not* included so the
    /// sweep can probe inside childless groups (tab bars, nav
    /// bars, toolbars) where hit-testable children hide.
    func leafFrames() -> [CGRect] {
        var out: [CGRect] = []
        collectLeafFrames(into: &out)
        return out
    }

    private func collectLeafFrames(into out: inout [CGRect]) {
        if isTerminal {
            out.append(CGRect(x: frameX, y: frameY,
                              width: frameWidth, height: frameHeight))
            return
        }
        for child in children { child.collectLeafFrames(into: &out) }
    }
}

// MARK: - Dedup + merge

extension AXNode {
    /// Stable identity for deduplication across the recursive walk
    /// and the grid sweep. Object identity is useless because the
    /// translator creates fresh translation objects per request.
    var fingerprint: String {
        let rx = frameX.rounded(), ry = frameY.rounded()
        let rw = frameWidth.rounded(), rh = frameHeight.rounded()
        return "\(role)|\(identifier ?? "")|\(label ?? "")|\(rx),\(ry),\(rw),\(rh)"
    }

    var center: (x: Double, y: Double) {
        (frameX + frameWidth / 2, frameY + frameHeight / 2)
    }

    /// Graft `extras` (from the grid sweep) into this tree. Each
    /// extra whose `fingerprint` isn't already in the tree is
    /// inserted under the **deepest existing container** whose
    /// frame contains its center. This keeps a discovered tab-bar
    /// button under its `AXGroup "Tab Bar"` parent — a flat append
    /// to the root would shadow it during client-side hit-tests.
    func merging(extras: [AXNode]) -> AXNode {
        var seen = Set<String>()
        collectFingerprints(into: &seen)
        let fresh = extras.filter { !seen.contains($0.fingerprint) }
        guard !fresh.isEmpty else { return self }
        // Remove duplicates within `fresh` itself.
        var deduped: [AXNode] = []
        for node in fresh {
            if seen.insert(node.fingerprint).inserted {
                deduped.append(node)
            }
        }
        return graft(deduped)
    }

    /// Recursively distribute `nodes` down to the deepest container
    /// that claims each one; unclaimed nodes attach to `self`.
    private func graft(_ nodes: [AXNode]) -> AXNode {
        var unclaimed = nodes
        let updatedChildren = children.map { child -> AXNode in
            let claimed = unclaimed.filter {
                child.containsPoint($0.center.x, $0.center.y)
            }
            if !claimed.isEmpty {
                unclaimed.removeAll {
                    child.containsPoint($0.center.x, $0.center.y)
                }
                return child.graft(claimed)
            }
            return child
        }
        return withChildren(updatedChildren + unclaimed)
    }

    private func collectFingerprints(into set: inout Set<String>) {
        set.insert(fingerprint)
        for child in children { child.collectFingerprints(into: &set) }
    }

    private func withChildren(_ newChildren: [AXNode]) -> AXNode {
        AXNode(
            role: role, subrole: subrole, label: label, value: value,
            identifier: identifier, title: title, help: help,
            frameX: frameX, frameY: frameY,
            frameWidth: frameWidth, frameHeight: frameHeight,
            enabled: enabled, focused: focused, hidden: hidden,
            children: newChildren
        )
    }
}
